guardian/typerighter
# Give each RuleMatch an explicit priority from Matchers, and order by priority when removing overlapping matches

Currently, style guide (regex) rules are removed in favour of dictionary rules when they overlap.

For example - let's say the text "Brigitte Trogneux" appears in an article, and the following are true:
- "Bridgitte" and "Trogneux" are absent from the dictionary
- "Bridgitte Trogneux" appears as a "Review" style guide rule

In this case, we would want "Bridgitte Trogneux" to appear with an orange 'Review' underline because of the style guide rule, but it would currently get a red 'Amend' from the dictionary matcher. This is because we are effectively prioritising by category id in reverse alphabetical order, which hasn't been a big issue until now because clashes between LanguageTool rules and style guide rules were not very common.

We want to make sure that Style Guide rules trump Dictionary rules.

**Note:** I've removed a test checking that rules are ordered by category, as it's no longer the case, and added a test elsewhere for overlapping rule removal where priority is specified.

## What does this change?

This PR adds an explicit priority to `CheckerRule` instances which can be used to deal with ranges of text that match more than one matcher type. Priority is associated with the rule type - specifically:

**LanguageTool rules are prioritised above Style Guide (regex) rules, which are prioritised above Dictionary rules.**

A bit of an oddity - Dictionary rules appear as `LtRule` when they are used after being extracted from the LanguageTool instance that contains them (part of the Dictionary Matcher) so we identify them there via their unique ID.

## How to test

1. Run this project locally alongside Composer, according to the instructions in the repos.
2. Make sure you have the following line in Composer's config in `~/.gu/flexible-composerbackend.properties` in order to use `local` Typerighter Checker rather than `CODE`: `typerighter.url=https://checker.typerighter.local.dev-gutools.co.uk`
3. Write "bridgitte is a grownup" on its own line and run a check. Does "bridgitte" get a red "Capitalization" category underline from the language tool "Amend" rule? Does "grownup" get an orange "review" underline from its regex "style guide" rule? If so, both rules are overriding collins "Amend" rules.
